### PR TITLE
Add Tests for HW

### PR DESCRIPTION
**Pull request by spencer-zepelin**
## In this PR:

- Updates to `.gitignore`
- Deletion of jupyter notebook checkpoints and removal from tracking
- Minor notebook formatting changes
- Problem 6.1 tests
- Problem 7.1 tests
- Problem 8.1 tests
- Test runners in each of those chapter directories (we could consider pulling them into a unified file or directory if that'd be cleaner)

### Notes:

- 8.2 is the monte carlo one and does not have deterministic results, so I'm not sure how we want to approach automated grading
- 7.2 is the password cracking problem and relies on external files
- 6.1 was as far as I've gotten, and I'll continue with 6.2 when I start up again.
- Might be overkill if we finish with the code for the course soon, but do we have any interest in trying out something like [](https://www.reviewnb.com/) so we can get good diffs on the PRs? We could probably get the academic license for free

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,5 @@
 .idea
+.tool-versions
+.envrc
```
Comment (spencer-zepelin) [inline .gitignore:3]: These are just for my local environment, I can pull them from the PR and just keep them locally if they bug ya

File: ch6/test_util.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+def fraction_equality(self, other):
+    return (
+        True
+        if self.numerator == other.numerator and self.denominator == other.denominator
+        else False
+    )
+
+
+def test61(examples):
```
Comment (chelseatroy) [inline ch6/test_util.py:9]: I'd like to push back on having a unique test runner for each individual problem. Maybe we need it and if we do that's fine, but the logic in a lot of these test methods looks very similar. My cursory assessment is that we could have a file in `/materials` with a small collection of test methods in it, which we could import as needed into each of the homework notebooks.

File: ch7/test_util.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+def test71(function, examples):
+    passed = 0
```
Comment (chelseatroy) [inline ch7/test_util.py:2]: Same comment above about an individual test runner for each problem

File: ch6/Homework_Problems.ipynb [hunk 4]
```diff
@@ -8,23 +8,69 @@
...
+    "The below should raise a `ZeroDivisionError`"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from problem_1 import Fraction\n",
```
Comment (chelseatroy) [inline ch6/Homework_Problems.ipynb:42]: Please put the file and the class (with a `pass`) in the directory so they're not running into an import error. See ch2 for examples of this.

Screen Shot 2021-01-08 at 8 59 19 AM

File: ch8/Homework_Problems.ipynb [hunk 5]
```diff
@@ -7,14 +7,49 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from test_util import testP1\n",
+    "from problem_1 import vectorize\n",
```
Comment (chelseatroy) [inline ch8/Homework_Problems.ipynb:34]: Please put the file and the class (with a `pass`) in the directory so they're not running into an import error (see ch2 for example: when they start from scratch running the ch2 homework problems notebook, they immediately see test failures.).

Screen Shot 2021-01-08 at 8 53 53 AM

**Review by chelseatroy — COMMENTED** [review]
Thanks for pulling out the checkpoints and cleaning things up! Much appreciated. Also saw that you left a couple notes for Lauren. That's perfect. 

I couldn't find a way to comment on those things inline, hence this overarching comment 😂 

Made a couple of notes inline amounting to two things:

1. Do you think we can pull the test running stuff into a common file at the root of the project? A lot of the logic looks _very_ similar. 
2. I'd like the problem files and tokens to already be there for them so when they run the test having not started the homework, they see test failures rather than import errors.

A sidenote on fraction equality: rather than us implement that in the test runner, I'd kinda rather change the question so _they_ implement that in `Fraction`. They're learning to implement dunder methods in that assignment, so I think it fits. I'm happy to make this change myself, though, if you prefer not to be changing homework requirements, haha

**Comment by chelseatroy** [comment]
I believe @lcampbe1 added the needed hw files so I'm gonna go ahead and merge this in, but definitely let's extract out those test runners :) In fact in this class students will do group activities focused on writing their _own_ test runner called phoenixtest. Naturally, I have bothered to draw a logo at this point, but not to actually plan any of the activities 😂

**Pull request merged**